### ISSUE TITLE
Regularisation and Soft Update Timing

### ISSUE THREAD

**Issue by beardyFace** [legalaspro/maddpg-zoo-torch#2]
This is the cleanest implementation of MADDPG that I have found, but I have a few questions about some functional differences from the original TensorFlow code (minor differences). I'm mostly curious about the reasoning, as various implementations of MADDPG do slightly different things to the original code as well, but don't really make it clear whether that was intentional. 

1) You omit the regularisation to the actor - is this deliberate?  

regularisation as in TF code
reg = (pred_action_i**2).mean() * 1e-3

actor_loss = -q_val.mean() + reg

2) You soft update the networks after updating all the agents - the original code soft updates after updating each individual agent. The paper doesn't make it clear which we should do - it is only from reading their code that you see them soft update per agent update. 

3) Recomputing the agents actions - the original code doesn't do this but I noticed you had a similar concept for recalculating the agents actions in the actor update. What was your reasoning for having that alternative? From the paper they should be "equal" but I could see an argument for it given the actors are changing - although they are meant to be "off-policy"

Compute actor loss
actions_pred = []
for i, agent in enumerate(self.agents):
    if i == agent_idx:
        actions_pred.append(current_agent.actor(states[i]))
    else: # Detach actions from other agents to prevent gradient flow
        # actions_pred.append(self.agents[i].actor(states[i]).detach())
        actions_pred.append(actions[i].detach())

Otherwise, your code has really helped narrow down my understanding of MADDPG thanks to clarity of the code compared to alternatives :)

**Comment by legalaspro** [comment]
Thanks a lot for the interest in my implementation and for the detailed questions! @beardyFace 

### 1. Actor Regularization Omission

I think have done it as a common simplification. I don’t remember it being used in the original DDPG paper, and in practice I’ve had better results just using gradient clipping (e.g. `norm = 1.0`) on the optimizer to deal with instability.

To even don't think about above issue, I’d actually suggest going straight to MATD3 based on TD3, where using two critics per agent and the clipped target largely removes the need for this kind of regularisation. [MATD3](https://github.com/legalaspro/rl-odyssey/blob/main/notebooks/multi-agent/2._MATD3-Continuous.ipynb) 

### 2. Soft Update Timing

The paper is ambiguous on soft update timing, and I guess the original TensorFlow code guess performs them per agent after each individual update - though I'd need to check the code to confirm. This in my opinion introduces "chaining" effect, where later agents see slightly more up-to-date target networks then earlier ones.

In my implementation I soft-update **once per training step after all agents have been update**. The idea is to keep things symmetric: all agents are trained against the same set of target networks for that cycle, which feels a bit cleaner in interdependent settings.

I think neither is strictly superior; it really depends on the environment and can be tested by shuffling the agent order. I've tried HAPPO and MAPPO—while chain-updated algorithms like HAPPO sound elegant in theory [Heterogeneous-Agent Reinforcement Learning](https://arxiv.org/abs/2304.09870) , MAPPO tended to work better for me despite the nice math of chained updates in HAPPO.

<img width="1956" height="1588" alt="Image" src="https://github.com/user-attachments/assets/d9097f37-3f4d-40bd-a46e-00f1fe331aa8" />

In the algorithm image, it looks like they do the soft update after running the update for each agent.

### 3. Recomputing Agents Actions

I believe the original code should recompute actions for the actor update. The critic receives actions from other actors as detached tensors, so gradients don't flow back to them— they're only computed for the current agent's actor. Only the path through agent i’s action should receive gradients; the other agents’ actions are treated as constants. I don't think the original code does it differently; this is how DDPG works overall. The actor gets updated by calculating its action, passing it to the critic, and then letting gradients flow back to adjust the original actor.

**Comment by beardyFace** [comment]
Hi,

Thanks for the response 

1 - Fair, my current plan was to faithfully recreate MADDPG just for understanding it - but yea next step seems like MATD3 or MASAC are just better regardless.

2- Yea - now you highlight the algorithm from the paper that is actually really clear on the target update timing. Interesting the their code appears to show otherwise...just one of those things I guess. I will try both and see if there is any meaningful difference otherwise will default to the paper setup.

3 - Yes, sorry to be clear, the current agent, you certainly recalculate the actions to get the gradient, etc. But for the other agents for the overall joint action, it appears to be unclear there -  specifically, these two lines in your code in the loop that creates the overall joint action.

The commented out part recomputes for each agent - with detach to remove gradients
actions_pred.append(self.agents[i].actor(states[i]).detach())

This follows the original code and just uses the actions from the buffer directly,
actions_pred.append(actions[i].detach())

Just curious what your thoughts are around that piece? I had presumed from the paper and their code that the second approach is fine - but I can see an argument to recompute for all the agents as well (updated to match the latest agents policy)

**Comment by legalaspro** [comment]
1 -  Totally agree – I also usually start by implementing the original algorithm

2 - My opinion is use "bathed" update mostly for symmetry, but testing both and defaulting to the papers is a way to go.

3 - You read my code exactly right. The commented line where I recompute all agents’ actions was from an initial coding: 
I switched to the replay-buffer actions for other agents because:
* For the current agent, we must recompute its action to get gradients.
* For the other agents, their actions are just treated as constants either way (detached).
* Using buffer actions keeps things strictly off-policy and avoids extra “chaining” through the current policies of other agents.

So that commented line is basically an artifact; the intended behavior in this implementation is to **only recompute the current agent’s action and use detached buffer actions for the others**.
